### PR TITLE
FP-FIX(frameworks): eliminate false positives across nextjs/server/tanstack/jotai/preact rules

### PR DESCRIPTION
**Pull request by aidenybai**
## FP-FIX — Framework adapter rules

Part of the false-positive hardening stack. Removes false positives across the framework adapters: **nextjs, server, tanstack-query, tanstack-start, jotai, preact, client, view-transitions**. The largest slice by file count. Based on #988 (foundation) — merge that first.

### Highlights
- **`nextjs-no-redirect-in-try-catch` / `tanstack-start-redirect-in-try-catch`** — a `redirect()`/`notFound()` thrown inside a `try` is only safe when the `catch` re-throws the **caught binding** (`throw e`). Fixed in this slice (see Bugbot below). Local `redirect` shadows and deferred-callback throws are not flagged.
- **`query-stable-query-client` / `query-no-query-in-effect` / `query-mutation-missing-invalidation` / `query-destructure-result`** — TanStack Query patterns narrowed to real misuse.
- **`jotai-select-atom-in-render-body` / `jotai-derived-atom-returns-fresh-object`** — event-handler / memoized callsites of `selectAtom` aren't flagged.
- **`server-*`** (fetch-revalidate, hoist-static-io, mutable-module-state, sequential-await) and **`nextjs-*`** (image sizes, inline-script id, no-a-element, client-fetch-for-server-data, client-side-redirect, font-link, script-in-head, side-effect-in-get-handler) tightened. Plus `tanstack-start-*`, `preact-no-children-length`, `client-*`, `no-document-start-view-transition`.

### Bugbot follow-ups
- **nextjs: any-throw suppresses** — FIXED: requires re-throwing the caught binding, so `throw new Error(...)` (which swallows the redirect) is reported again.
- **tanstack-start: `isRedirect()` call alone suppresses** — FIXED: requires an actual re-throw, not just a guard call.
- **jotai handler-name / query inner-helper called during render** — declined: closing these requires call-graph analysis (proving a `handle*`-named fn / lowercase helper actually runs during render), which would reintroduce FPs on the common deferred-handler case. (See threads.)

### Eval results

| Check | Result |
| --- | --- |
| Repos scanned | 500 distinct (deduped from repos.json's 8,423 entries) |
| RootDir scans | 467 succeeded / 33 worker failures (giant repos: sentry, appsmith, …) |
| Target rules | all 34 changed on this branch |
| Diagnostics | 1,603 across 21 firing rules |
| Manually classified | 161 hits (≤10 distinct-repo samples per rule) |
| False positives found | 6 (3.7%) → 2 patterns fixed in `407fdbf5` w/ regression tests, rest documented |
| Caching | disabled (`REACT_DOCTOR_NO_CACHE=1` + fresh RDE spec cache) |
| Output artifact | `.evals/-Users-…-prague.jsonl` (RDE), filtered hits in eval comment |

Zero-hit audit: the 13 silent rules (incl. all 7 tanstack-start) were probed directly — in-scope corpus code was genuinely clean (every `throw redirect` an idiomatic `beforeLoad` guard), and CLI fixtures confirmed each hardened rule still fires on violating shapes while allowing the documented-safe ones. Details in the [eval comment](https://github.com/millionco/react-doctor/pull/995#issuecomment-4859798390).

### Test plan
- [x] framework unit + regression suites green
- [x] nextjs redirect e2e fixture still triggers under the stricter rule
- [x] typecheck clean
- [x] RDE eval: 500 distinct OSS repos, 161 hits classified, both systematic FPs fixed
- [ ] Full CI after foundation merges + retarget to `main`

---

> [!NOTE]
> **Low Risk**
> Changes are confined to static lint heuristics and regression tests in the plugin; they do not alter application runtime behavior. Residual risk is misclassified diagnostics if new edge cases slip past the expanded analysis.
> 
> **Overview**
> **Patch release** for `oxlint-plugin-react-doctor` that tightens framework adapter rules (Next.js, server, TanStack Query, Jotai, Preact, client, view transitions) so legitimate patterns stop firing while real violations still report.
> 
> **Redirect / try-catch** — `nextjs-no-redirect-in-try-catch` (and the parallel TanStack Start rule) now resolves real `next/navigation` imports, ignores shadowed locals, and uses `findGuardingTryStatement` so catches that re-throw the caught error, bare `try/finally`, and deferred callbacks are not treated as swallowing; `throw new Error` in catch still flags.
> 
> **Effects & mounts** — Client fetch/redirect-in-`useEffect` rules follow IIFEs, immediately invoked locals, and promise chains run on mount, but skip event handlers, cleanup, and later-fired callbacks via `collectEffectInvokedFunctions` / `containsFetchCall` boundaries.
> 
> **Server** — `server-hoist-static-io` taints request-derived paths through destructuring and bindings; `server-no-mutable-module-state` only reports `const` containers that are actually mutated (aliases, nested writes, one-hop same-file helpers); `server-fetch-without-revalidate` abstains on non-literal or spread options; `server-sequential-independent-await` ignores auth/connect gates and already-started promises; GET handler side-effect detection follows one same-file helper hop and no longer flags read-only “mutating” route names alone.
> 
> **TanStack Query / Jotai** — Invalidation recognizes scoped `invalidateQueries` and tRPC `use*Utils` → `.invalidate()`; destructure-result distinguishes forwarding the whole query from field reads; raw `*QueryAtom` imports require envelope-shaped consumption; `selectAtom` / derived-atom rules add memo, effect, handler, and reference-stability escape hatches.
> 
> **Next.js & client UX** — Image/script rules respect JSX spreads; anchors skip `//`, downloads, and `_blank`; passive listeners resolve handlers for `preventDefault` and explicit `passive`; localStorage versioning accepts camelCase `V2` keys; Preact `children` checks gate on component-like functions.
> 
> Regression tests were added across these rules; changeset notes validation on **500 OSS repos**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5b1b8799940e9567d95a67a4a4a248ecc525b688. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-redirect-in-try-catch.ts [hunk 1]
```diff
@@ -1,9 +1,67 @@
...
+      child !== handler.body &&
+      (isNodeOfType(child, "ArrowFunctionExpression") ||
+        isNodeOfType(child, "FunctionExpression") ||
+        isNodeOfType(child, "FunctionDeclaration"))
+    ) {
+      return false;
+    }
+    if (isNodeOfType(child, "ThrowStatement")) {
+      didRethrow = true;
+      return false;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-redirect-in-try-catch.ts]: Fixed in 7e8baa66. The nextjs catch check now requires re-throwing the CAUGHT binding (`throw e`), matching the documented `if (isRedirectError(e)) throw e` pattern. A catch that throws a fresh error (`throw new Error(...)`) — which swallows the redirect's control-flow error — is reported again. Regression test added.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-redirect-in-try-catch.ts [hunk 2]
```diff
@@ -1,9 +1,67 @@
...
+      forwards = true;
+      return false;
+    }
+    if (
+      isNodeOfType(child, "CallExpression") &&
+      isNodeOfType(child.callee, "Identifier") &&
+      REDIRECT_CONTROL_FLOW_GUARDS.has(child.callee.name)
+    ) {
+      forwards = true;
+      return false;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-redirect-in-try-catch.ts]: Fixed in 7e8baa66. `catchClauseForwardsRedirect` no longer treats a bare `isRedirect()`/`isNotFound()` call as forwarding; it requires an actual re-throw of the caught binding. A catch that checks the guard but only logs is reported again. Regression test added.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/jotai/jotai-select-atom-in-render-body.ts [hunk 3]
```diff
@@ -38,15 +46,124 @@ const isImportedSelectAtom = (callExpression: EsTreeNodeOfType<"CallExpression">
...
+  ) {
+    return true;
+  }
+
+  if (
+    isNodeOfType(functionNode, "FunctionDeclaration") &&
+    functionNode.id &&
+    HANDLER_FUNCTION_NAME_PATTERN.test(functionNode.id.name)
+  ) {
+    return true;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/jotai/jotai-select-atom-in-render-body.ts]: Declining — closing this would require call-graph analysis. `isDeferredCallback` uses a name-based heuristic (`handle*`/`on*`) because the overwhelmingly common case is a real event handler; flagging a `handle*`-named function that *happens* to be invoked synchronously during render means proving render-time invocation, which needs a call graph. Adding that risks false positives on ordinary handlers (the exact pattern this exemption protects). A handler-named function actually called during render is a rare anti-pattern; accepting that narrow FN to keep the common case FP-free is the right tradeoff here.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-query/query-stable-query-client.ts [hunk 4]
```diff
@@ -15,70 +34,29 @@ export const queryStableQueryClient = defineRule({
...
-          !isNodeOfType(node.callee, "Identifier") ||
-          node.callee.name !== TANSTACK_QUERY_CLIENT_CLASS
-        )
+      // Only fire when the nearest enclosing function is the component itself
+      // — i.e. the construction runs in the render body. A nested closure
+      // (event handler, stable-hook initializer) defers it, so it's stable.
+      let cursor: EsTreeNode | null | undefined = node.parent;
+      while (cursor) {
+        if (isFunctionLike(cursor)) {
+          if (!isComponentFunction(cursor)) return;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-query/query-stable-query-client.ts]: Declining — same class as the jotai comment: this needs call-graph analysis. The rule stops at the nearest enclosing function; `new QueryClient()` inside a lowercase helper that the component happens to call during render is missed because proving the helper runs during render (vs. being an event handler / deferred callback) requires tracing call sites. Doing that imprecisely would reintroduce false positives on helpers that are NOT called during render. The component/hook-scope heuristic is the deliberate FP-averse boundary; the nested-helper case is an accepted narrow miss.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-client-fetch-for-server-data.ts [hunk 5]
```diff
@@ -28,7 +28,7 @@ export const nextjsNoClientFetchForServerData = defineRule({
         if (!fileHasUseClient || !isHookCall(node, EFFECT_HOOK_NAMES)) return;
 
         const callback = getEffectCallback(node);
-        if (!callback || !containsFetchCall(callback)) return;
+        if (!callback || !containsFetchCall(callback, { stopAtFunctionBoundary: true })) return;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-client-fetch-for-server-data.ts:31]: Declining — `stopAtFunctionBoundary: true` is intentional. Descending into nested function bodies would also flag `fetch` inside event handlers / deferred callbacks (which are correct client fetches), reintroducing false positives. Reliably distinguishing a mount-time async IIFE in `useEffect` from a deferred handler needs call-context analysis beyond this rule's scope; the boundary is the FP-averse choice, and the IIFE-in-effect case is an accepted miss.

**Comment by rayhanadev** [comment]
Pushed a follow-up **deslop / consolidation** commit (`a246ff1d`) on top of the FP fixes. The hardening pass had copy-pasted the same helpers across many rules, so this collapses them:

- 4 new shared utils — `find-enclosing-function`, `get-function-binding-name`, `collect-handler-referenced-names`, `catch-clause-rethrows-caught` — plus a `HOOK_NAME_PATTERN` constant, replacing inline copies across ~6 / 4 / 2 / 2 rules (`collect-handler-referenced-names` was a verbatim jotai↔navigate port).
- Reuse existing utils (`isFunctionLike`, `getCalleeName`, `findJsxAttribute`) in place of re-implemented inline copies; dropped one nested ternary.

Behaviour-preserving: full plugin suite unchanged (**7409 passing / 88 skipped**), typecheck + lint + format clean, `react-doctor --staged` 100/100. Net **−208 lines** (+194 / −402 across 19 files).

**Comment by rayhanadev** [comment]
## RDE eval results (500 distinct OSS repos, cache disabled)

| Check | Result |
| --- | --- |
| Repos scanned | 500 distinct (deduped from repos.json) |
| RootDir scans | 467 succeeded, 33 worker failures (giant repos: sentry, appsmith, …) |
| Target rules | all 34 changed on this branch |
| Diagnostics | 1,603 across 21 firing rules |
| Manually classified | 161 hits (≤10 distinct-repo samples/rule) |
| False positives | 6 confirmed (3.7%) → 2 patterns fixed in `407fdbf5`, 1 deferred |
| Artifacts | `.evals/-Users-…-prague.jsonl`, `/tmp/rde-target-hits.json` |

**Fixed from eval findings** (with regression tests):
- `server-hoist-static-io`: request-derived paths through intermediate bindings (`params → pathArray → filePath → fullPath`) no longer flagged — taint now propagates through declarations (3/10 sampled hits were this FP).
- `query-mutation-missing-invalidation`: tRPC's `utils.x.invalidate()` now counts as cache invalidation.

**Deferred** (documented limits): helper-wrapped invalidation (`invalidateDashboardQueries(queryClient)`) needs same-file call resolution; `nextjs-no-side-effect-in-get-handler` can flag method-branched shared handlers (`export { handle as GET, handle as POST }`).

**Zero-hit audit**: 13 silent rules verified — corpus route files in scanned rootDirs contained no violations (all `throw redirect` sites were idiomatic `beforeLoad` guards), and CLI probes confirmed the hardened rules still fire on violating shapes (bare-catch redirect swallow → flags; `isRedirectError` rethrow → silent, which the old rule false-positived).

Suite: 7412 passed / 88 skipped, typecheck + format clean.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-query/query-destructure-result.ts [hunk 6]
```diff
@@ -1,18 +1,39 @@
...
+  let forwarded = false;
+  walkAst(scope, (node: EsTreeNode) => {
+    if (forwarded) return;
+    if (!isNodeOfType(node, "Identifier") || node.name !== bindingName) return;
+    if (node === declarator.id) return;
+    const parent = node.parent;
+    if (isNodeOfType(parent, "MemberExpression") && parent.object === node) return;
+    forwarded = true;
+  });
+  return forwarded;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-query/query-destructure-result.ts]: Fixed in e7c23b08 — `isForwardedBinding` now positively matches the documented forwarding shapes (returned incl. implicit/tuple return, call argument, JSX, spread, re-bound) instead of treating any non-member mention as forwarding. A bare dependency-array mention (`useEffect(..., [query])`) no longer suppresses; regression tests cover the dep-array, later-destructure, and tuple-return shapes.

**Review by rayhanadev — APPROVED** [review]

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-redirect-in-try-catch.ts [hunk 7]
```diff
@@ -12,37 +32,21 @@ export const tanstackStartRedirectInTryCatch = defineRule({
...
-          node,
-          message: `throw ${argument.callee.name}() inside a try block gets swallowed, so the redirect silently fails.`,
-        });
-      },
-    };
-  },
+      context.report({
+        node,
+        message: `throw ${argument.callee.name}() inside a try block gets swallowed, so the redirect silently fails.`,
+      });
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-redirect-in-try-catch.ts:32]: Fixed in a114b910 — the tanstack rule now uses the same walk the nextjs sibling already had, extracted as the shared `findGuardingTryStatement` util: a try only counts as swallowing when its BLOCK contains the throw AND it has a catch handler. Bare `try/finally` and throws inside `finally` no longer flag, and the walk keeps climbing past catch/finally so a throw in an inner catch swallowed by an OUTER try/catch is now correctly reported (a false negative the old early-return also had). Regression tests cover all three shapes.
